src/presence/HeartbeatComponent.js

```diff
@@ -0,0 +1,63 @@
+// @flow strict-local
+import { PureComponent } from 'react';
+import { AppState } from 'react-native';
+import type { Auth, Dispatch } from '../types';
+
+import { connect } from '../react-redux';
+import { tryGetAuth } from '../account/accountsSelectors';
+import { reportPresence } from '../actions';
+import Heartbeat from './heartbeat';
+
+type Props = $ReadOnly<{|
+  dispatch: Dispatch,
+  auth: Auth | void,
+|}>;
+
+/**
+ * Component providing a recurrent `presence` signal.
+ */
+// Note that "PureComponent" is of questionable veracity: this component's
+// entire purpose is to emit network calls for their observable side effects.
+// However, it is at least true that there is never a need to call `render()` if
+// the props haven't changed.
+class PresenceHeartbeat extends PureComponent<Props> {
```

Comment:
I think that's actually 100% consistent with what `PureComponent` is intended to mean:
https://reactjs.org/docs/react-api.html#reactpurecomponent

After all, almost all components are meant to cause something to appear in the UI; and any component with a widget like a button (that does something) is capable of causing quite general side effects, often a network request. All of that is independent of being a `PureComponent`. So the sense of "pure" in `PureComponent` has to be a lot narrower than "this is side-effect-free code".

I think in fact it really runs only in the dual direction: it says it's free of *suffering* certain kinds of side effects, but nothing about what side effects it may *cause*.

It could perhaps have a more lucid name than it does. :slightly_smiling_face:

Reply:
> After all, almost all components are meant to cause something to appear in the UI; and any component with a widget like a button (that does something) is capable of causing quite general side effects, often a network request.

This is true, as far as it goes – but `PureComponent.render()` in particular is expected to do precisely none of those things, and instead only to return a specification of how and when one might later go about doing them.

That expectation is flagrantly violated by this component, whose `.render()` can, and often will, fire off a network request before returning. It's less pure than many `Component.render()` functions. The kindest thing one can say about it, I think, is that it is idempotent even when considered as a function from `World` to `World`.

(Well, modulo modifications of `this` and its prototype chain. But if you're breaking the equivalence of `this.foo.bind(this)` and `() => this.foo()`, then nothing holds anyway.)

Reply:
> That expectation is flagrantly violated by this component, whose `.render()` can, and often will, fire off a network request before returning.

Hmm, I see.

I think that's a sign that `render` isn't the right place to put that call -- quite independently of `PureComponent` vs. its base class `Component`.

Consulting the API docs:
https://reactjs.org/docs/react-component.html
I think `componentDidUpdate` is the intended place to put this kind of call. That's based on its own description, and also on "use this instead" pointers to it in the descriptions of `getDerivedStateFromProps` and the deprecated `UNSAFE_componentWillReceiveProps`.

Reply:
:+1:

(The docs aren't explicit as to whether `componentDidUpdate` will fire if the `post-render()` reconciliation phase says there are no changes, but [it seems that it will](https://github.com/reactjs/reactjs.org/pull/1230).)

Reply:
Oof, that's a regrettable form for documentation to be in. :-/ Glad you tracked it down -- please mention it somewhere so that we, at least, can hang onto that information.

Reply:
It is now the case that this component performs side effects as a side effect of `render()`, rather than in `render()` directly.

Modulo that change, I think this warning remains appropriate.